flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/TableStore.java

```diff
@@ -171,6 +181,83 @@ private FileStore buildFileStore() {
                 tableIdentifier, options, user, partitionType, keyType, valueType, mergeFunction);
     }
 
+    // --------------------------------------------------------------------------------------------
+
+    /** A builder for constructing a immutable {@link TableStore}. */
+    public static final class TableStoreBuilder {
+
+        private Configuration options;
+        private ObjectIdentifier tableIdentifier;
+        private RowType type;
+
+        private List<String> primaryKeys = new ArrayList<>();
+        private List<String> partitionKeys = new ArrayList<>();
+        private String user = UUID.randomUUID().toString();
+
+        public TableStoreBuilder withConfiguration(Configuration options) {
+            this.options = options;
+            return this;
+        }
+
+        public TableStoreBuilder withTableIdentifier(ObjectIdentifier tableIdentifier) {
+            this.tableIdentifier = tableIdentifier;
+            return this;
+        }
+
+        public TableStoreBuilder withSchema(RowType type) {
+            this.type = type;
+            return this;
+        }
+
+        public TableStoreBuilder withPrimaryKeys(List<String> primaryKeys) {
+            this.primaryKeys = primaryKeys;
+            return this;
+        }
+
+        public TableStoreBuilder withPartitionKeys(List<String> partitionKeys) {
+            this.partitionKeys = partitionKeys;
+            return this;
+        }
+
+        public TableStoreBuilder withUser(String user) {
+            this.user = user;
+            return this;
+        }
+
+        public TableStore build() {
+            List<int[]> indices = adjustAndValidate();
+            return new TableStore(
+                    options, tableIdentifier, type, indices.get(0), indices.get(1), user);
+        }
+
+        private List<int[]> adjustAndValidate() {
```

Comment:
A simpler approach would be:
```
       public TableStore withPartitions(int[] partitions) {
        this.partitions = partitions;
        adjustAndValidateKeys();
        return this;
    }

    public TableStore withPrimaryKeys(int[] primaryKeys) {
        this.primaryKeys = primaryKeys;
        adjustAndValidateKeys();
        return this;
    }
    public void adjustAndValidateKeys() {
        // ... do validate and adjust
    }
```

Reply:
I don't think so. Your approach has put an assumption that `#withSchema` is called first before `#withPartitions` and `#withPrimaryKeys`.
What if `new TableStore(options).withPrimaryKeys(pk).withSchema(rowtype)` is called? 

In my opinion, if we put the index adjusting and validation within the `TableStore`, it must be computed at the phase before `TableStore` is fully instantiated, i.e. within some `#build()`

Reply:
Why we need schema? We can just show index in exception without schema.

Reply:
> Why we need schema? We can just show index in exception without schema.

because index adjusting needs to be computed based on the physical schema. If we throw an exception, does this approach against that the builder pattern doesn't impose call order? How could we call this a builder pattern?

Reply:
It's weird that `TableStore` must be first specified with schema then specified with pk and partitions, why do we impose the order?

Reply:
I don't get it.
In my understanding, this is only index adjustment, why must need schema.

Reply:
> I don't get it. In my understanding, this is only index adjustment, why must need schema.

So in your understanding, index adjustment is based on what?  Below is the current index adjusting in `TableStoreFactory`

```java
primaryKeys =
                    schema.getPrimaryKey().get().getColumns().stream()
                            .mapToInt(rowType.getFieldNames()::indexOf)
                            .toArray()
```

without schema, how could `mapToInt` be accomplished?